californium-core/src/main/java/org/eclipse/californium/core/network/stack/ReliabilityLayer.java

```diff
@@ -235,6 +237,11 @@ public void receiveRequest(final Exchange exchange, final Request request) {
 							sendResponse(exchange, currentResponse);
 						}
 						return;
+					} else if (currentResponse.isNotification()) {
+						// notifications are kept in the exchange store, so
+						// prepare retransmission counter for retransmission
+						int failedCount = exchange.getFailedTransmissionCount() + 1;
+						exchange.setFailedTransmissionCount(failedCount);
```

Comment:
I don't get the point here. 
`FailedTransmissountCount` is a best effort counter which tries to count how many time we failed to send a message for this exchange ?

And as we are able to count this even for NON notification we try to do it ? 
I try to read the test to understand the use case but still don't get it :grimacing:

Reply:
The exchange store keeps only messages, when a answer from the other peer is considered. this is mainly CON and for notifies also NON to react on RST to cancel the observation.

[Check NON with MID collision](https://github.com/eclipse/californium/blob/master/californium-core/src/main/java/org/eclipse/californium/core/network/InMemoryMessageExchangeStore.java#L259)

That check takes the `FailedTransmissountCount` to check, if this is a failure or a retransmission. Therefore I increment this now also for non notifies retransmission.

I added two test, one with piggy-backed response, which would work without this fix. And one with separate response, which is affected by this bug.
If the request gets retransmitted, the duplicate will cause first the ACK (as on the first receive of that request), and then the NON notify , which was already send and so was already stores in the exchange store. With this fix, this is now handled as retransmission an so the check is passed.

Reply:
a bit clearer to me, thx